### PR TITLE
Wip

### PR DESCRIPTION
**Pull request by purplefox**

### CODE REVIEW

**Comment by purplefox** [comment]
@alexlehm if you are around some time soon to discuss this please ping me.

**Comment by alexlehm** [comment]
Ok, please do a new review of the PR, I have fixed the tests taking long and I have added to the documentation

File: vertx-mail-client/src/main/asciidoc/js/index.adoc [hunk 1]
```diff
@@ -1,70 +1,166 @@
-= Vert.x Mail (SMTP client implementation)
+= Vert.x Mail client (SMTP client implementation)
 
-`link:jsdoc/mail_client-MailClient.html[MailClient]` simple example:
+Vert.x client for sending SMTP emails via a local mail server (e.g. postfix),
+by external mail server (e.g. googlemail or aol) or by the vert.x event bus via
+a mail service running on another machine on the local network.
```
Comment (purplefox) [inline vertx-mail-client/src/main/asciidoc/js/index.adoc]: don't mention the mail service in the mail client docs

File: vertx-mail-client/src/main/asciidoc/js/index.adoc [hunk 2]
```diff
@@ -1,70 +1,166 @@
...
+
+The client supports a few additional auth methods like DIGEST-MD5 and has full
+support for TLS and SSL and is completely asynchronous. The client supports
+connection pooling to keep connections open for an specific time to be reused.
+
+== Creating a client
+
+There are two possibilities to send mails, either by creating a client that
+opens SMTP connections from the local jvm or by creating a service that
```
Comment (purplefox) [inline vertx-mail-client/src/main/asciidoc/js/index.adoc]: again, don't mention the service - that belongs in the service docs

File: vertx-mail-client/src/main/asciidoc/js/index.adoc [hunk 3]
```diff
@@ -1,70 +1,166 @@
...
 
-mailClient.sendMail(email, function (result, result_err) {
-  if (result_err == null) {
-    console.log(result);
-  } else {
-    console.log("got exception");
-    result_err.printStackTrace();
-  };
-});
+For documentation about the MailService interface, please take a look at the documentation
```
Comment (purplefox) [inline vertx-mail-client/src/main/asciidoc/js/index.adoc]: Again, don't mention the service.

File: vertx-mail-client/src/main/asciidoc/js/index.adoc [hunk 4]
```diff
@@ -1,70 +1,166 @@
...
 });
 
-----
\ No newline at end of file
+----
+
+== Mail-client data objects
+
+The objects used by the MailClient service are all @DataObject classes, which means that they have a few properties
+and can be converted from and to JSON. In Java the regular getter and setter methods can be used for the properties
```
Comment (purplefox) [inline vertx-mail-client/src/main/asciidoc/js/index.adoc]: Don't mention other languages.

File: vertx-mail-client/src/main/java/io/vertx/ext/mail/MailConfig.java [hunk 5]
```diff
@@ -71,20 +71,35 @@
   /**
    * use this hostname for HELO/EHLO command
    */
-  private String ehloHostname;
+  private String ownHostname;
+
+  /**
```
Comment (purplefox) [inline vertx-mail-client/src/main/java/io/vertx/ext/mail/MailConfig.java]: Don't javadoc private methods

**Comment by purplefox** [comment]
Alex, I think this is almost there, some general points:

1. The code to share the client is missing (see the createShared/createNonShared on the other clients)
2. I don't think we should be letting the user supply a netClientOptions
3. Is IDLE timeout really necessary? We don't use this on other clients and it seems to add extra complexity
4. There are lots of TODOs in the code. If they are really TODOs please add issues for them, otherwise delete them from the code.

**Comment by purplefox** [comment]
Perhaps we should leave out the shared stuff for now.....

I'm going to merge as is. But please lets address the other issues.

**Pull request merged**